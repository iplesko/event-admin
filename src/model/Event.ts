export interface Event {
  id?: number,
  from: Date,
  to: Date,
  name: string,
  description: string,
}
