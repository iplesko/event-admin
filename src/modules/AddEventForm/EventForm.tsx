import React, { ChangeEvent, FormEvent, ReactElement, useEffect, useState } from 'react';
import { Button, FormGroup, Input, Label, Modal, ModalBody, ModalFooter, ModalHeader } from 'reactstrap';
import { Event } from '../../model/Event'
import formatDate from "../../function/formatDate";

interface Props {
  heading: string;
  buttonText: string;
  onConfirm: (e: Event) => void;
  visible: boolean;
  toggleVisibility: () => void;
  displayedEvent?: Event;
}

const empty = () => ({
  name: '', from: new Date(), to: new Date(), description: '',
});

export default ({ heading, buttonText, onConfirm, visible, toggleVisibility, displayedEvent }: Props): ReactElement => {
  const [ event, setEvent ] = useState<Event>({
    name: '', from: new Date(), to: new Date(), description: '',
  });

  const handleChangeText = (e: ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.target;
    setEvent(prevState => ({ ...prevState, [id]: value }));
  }

  const handleChangeDate = (e: ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.target;
    setEvent(prevState => ({ ...prevState, [id]: new Date(value) }));
  }

  const confirm = (e: FormEvent) => {
    onConfirm(event);
    toggleVisibility();
    e.preventDefault();
  }

  useEffect(() => {
    setEvent(displayedEvent || empty());
  }, [ visible ]);

  return (
      <div>
        <Modal isOpen={visible} toggle={toggleVisibility}>
          <form onSubmit={confirm}>
            <ModalHeader toggle={toggleVisibility}>{heading}</ModalHeader>
            <ModalBody>

              <FormGroup required>
                <Label for="name">Event name *</Label>
                <Input required type="text" id="name" value={event.name} onChange={handleChangeText}/>
              </FormGroup>

              <FormGroup required>
                <Label for="from">From *</Label>
                <Input required type="date" id="from" value={formatDate(event.from)} onChange={handleChangeDate}/>
              </FormGroup>

              <FormGroup required>
                <Label for="to">To *</Label>
                <Input required type="date" id="to" value={formatDate(event.to)} onChange={handleChangeDate}/>
              </FormGroup>

              <FormGroup>
                <Label for="description">Description</Label>
                <Input type="textarea" id="description" value={event.description} onChange={handleChangeText}/>
              </FormGroup>

            </ModalBody>
            <ModalFooter>
              <Button color="primary" type="submit">{buttonText}</Button>{' '}
              <Button color="secondary" onClick={toggleVisibility}>Cancel</Button>
            </ModalFooter>
          </form>
        </Modal>
      </div>
  );
}
