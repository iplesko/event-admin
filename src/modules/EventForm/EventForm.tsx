import React, {
  ChangeEvent, FormEvent, ReactElement, useEffect, useState,
} from 'react';
import {
  Button, FormGroup, Input, Label, Modal, ModalBody, ModalFooter, ModalHeader,
} from 'reactstrap';
import { Event } from '../../model/Event';
import formatDate from '../../function/formatDate';

interface Props {
  heading: string;
  buttonText: string;
  onConfirm: (e: Event) => void;
  visible: boolean;
  toggleVisibility: () => void;
  displayedEvent?: Event;
}

const empty = () => ({
  name: '', from: new Date(), to: new Date(), description: '',
});

export default ({
  heading, buttonText, onConfirm, visible, toggleVisibility, displayedEvent,
}: Props): ReactElement => {
  const [event, setEvent] = useState<Event>(empty());

  const handleChangeText = (e: ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.target;
    setEvent((prevState) => ({ ...prevState, [id]: value }));
  };

  const handleChangeDateRange = (e: ChangeEvent<HTMLInputElement>) => {
    let from: Date;
    let to: Date;

    const { id, value } = e.target;
    if (id === 'to') {
      to = new Date(value);
      from = event.from > to ? to : event.from;
    } else {
      from = new Date(value);
      to = from > event.to ? from : event.to;
    }
    setEvent((prevState) => ({ ...prevState, from, to }));
  };

  const confirm = (e: FormEvent) => {
    onConfirm(event);
    toggleVisibility();
    e.preventDefault();
  };

  useEffect(() => {
    if (visible) {
      setEvent(displayedEvent || empty());
    }
  }, [visible]);

  return (
    <Modal isOpen={visible} toggle={toggleVisibility}>
      <form onSubmit={confirm}>
        <ModalHeader toggle={toggleVisibility}>{heading}</ModalHeader>
        <ModalBody>

          <FormGroup required>
            <Label for="name">Event name *</Label>
            <Input required type="text" id="name" value={event.name} onChange={handleChangeText} />
          </FormGroup>

          <FormGroup required>
            <Label for="from">From *</Label>
            <Input required type="date" id="from" value={formatDate(event.from)} onChange={handleChangeDateRange} />
          </FormGroup>

          <FormGroup required>
            <Label for="to">To *</Label>
            <Input required type="date" id="to" value={formatDate(event.to)} onChange={handleChangeDateRange} />
          </FormGroup>

          <FormGroup>
            <Label for="description">Description</Label>
            <Input type="textarea" id="description" value={event.description} onChange={handleChangeText} />
          </FormGroup>

        </ModalBody>
        <ModalFooter>
          <Button color="primary" type="submit">{buttonText}</Button>
          {' '}
          <Button color="secondary" onClick={toggleVisibility}>Cancel</Button>
        </ModalFooter>
      </form>
    </Modal>
  );
};
