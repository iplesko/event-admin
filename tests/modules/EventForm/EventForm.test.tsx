import React from 'react';
import { shallow } from 'enzyme';
import EventForm from '../../../src/modules/EventForm';
import { Event } from '../../../src/model/Event';

describe('EventForm component', () => {
  let testEvent: Event;

  beforeEach(() => {
    testEvent = {
      description: 'Test event description',
      from: new Date(2020, 3, 4),
      id: 123,
      name: 'Test event name',
      to: new Date(2020, 3, 6),
    };
  });

  it('should render correctly', () => {
    const wrapper = shallow(<EventForm
      displayedEvent={testEvent}
      toggleVisibility={jest.fn()}
      visible
      buttonText="Test button text"
      heading="Test heading"
      onConfirm={jest.fn()}
    />);

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should call callback and close modal on confirm', () => {
    const callbackMock = jest.fn();
    const toggleVisibilityMock = jest.fn();

    const wrapper = shallow(<EventForm
      displayedEvent={testEvent}
      toggleVisibility={toggleVisibilityMock}
      visible
      buttonText="Test button text"
      heading="Test heading"
      onConfirm={callbackMock}
    />);

    wrapper.find('Input[id="name"]').simulate('change', { target: { id: 'name', value: 'changed name' } });
    wrapper.find('Input[id="from"]').simulate('change', { target: { id: 'from', value: '2021-05-06' } });
    wrapper.find('Input[id="to"]').simulate('change', { target: { id: 'to', value: '2021-05-07' } });
    wrapper.find('Input[id="description"]').simulate('change', { target: { id: 'description', value: 'changed description' } });

    const preventDefaultMock = jest.fn();
    wrapper.find('form').simulate('submit', { preventDefault: preventDefaultMock });

    expect(callbackMock).toHaveBeenCalledWith({
      name: 'changed name',
      from: new Date('2021-05-06'),
      to: new Date('2021-05-07'),
      description: 'changed description',
    } as Event);
    expect(toggleVisibilityMock).toHaveBeenCalled();
    expect(preventDefaultMock).toHaveBeenCalled();
  });

  it('should correct from date, when to date is set', () => {
    const wrapper = shallow(<EventForm
      toggleVisibility={jest.fn()}
      visible
      buttonText="Test button text"
      heading="Test heading"
      onConfirm={jest.fn()}
    />);

    wrapper.find('Input[id="from"]').simulate('change', { target: { id: 'from', value: '2020-05-06' } });
    wrapper.find('Input[id="to"]').simulate('change', { target: { id: 'to', value: '2020-03-31' } });

    const from = wrapper.find('Input[id="from"]').prop('value');
    expect(from).toBe('2020-03-31');
  });

  it('should correct to date, when from date is set', () => {
    const wrapper = shallow(<EventForm
      toggleVisibility={jest.fn()}
      visible
      buttonText="Test button text"
      heading="Test heading"
      onConfirm={jest.fn()}
    />);

    wrapper.find('Input[id="to"]').simulate('change', { target: { id: 'to', value: '2020-05-06' } });
    wrapper.find('Input[id="from"]').simulate('change', { target: { id: 'from', value: '2020-10-20' } });

    const from = wrapper.find('Input[id="to"]').prop('value');
    expect(from).toBe('2020-10-20');
  });
});
