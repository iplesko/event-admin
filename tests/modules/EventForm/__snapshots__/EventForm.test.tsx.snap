// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventForm component should render correctly 1`] = `
"<Modal isOpen={true} toggle={[Function: mockConstructor]} autoFocus={true} centered={false} scrollable={false} role=\\"dialog\\" backdrop={true} keyboard={true} zIndex={1050} fade={true} onOpened={[Function: noop]} onClosed={[Function: noop]} modalTransition={{...}} backdropTransition={{...}} unmountOnClose={true} returnFocusAfterClose={true} container=\\"body\\">
  <form onSubmit={[Function: confirm]}>
    <ModalHeader toggle={[Function: mockConstructor]} tag=\\"h5\\" wrapTag=\\"div\\" closeAriaLabel=\\"Close\\" charCode={215}>
      Test heading
    </ModalHeader>
    <ModalBody tag=\\"div\\">
      <FormGroup required={true} tag=\\"div\\">
        <Label for=\\"name\\" tag=\\"label\\" widths={{...}}>
          Event name *
        </Label>
        <Input required={true} type=\\"text\\" id=\\"name\\" value=\\"\\" onChange={[Function: handleChangeText]} />
      </FormGroup>
      <FormGroup required={true} tag=\\"div\\">
        <Label for=\\"from\\" tag=\\"label\\" widths={{...}}>
          From *
        </Label>
        <Input required={true} type=\\"date\\" id=\\"from\\" value=\\"2020-08-24\\" onChange={[Function: handleChangeDateRange]} />
      </FormGroup>
      <FormGroup required={true} tag=\\"div\\">
        <Label for=\\"to\\" tag=\\"label\\" widths={{...}}>
          To *
        </Label>
        <Input required={true} type=\\"date\\" id=\\"to\\" value=\\"2020-08-24\\" onChange={[Function: handleChangeDateRange]} />
      </FormGroup>
      <FormGroup tag=\\"div\\">
        <Label for=\\"description\\" tag=\\"label\\" widths={{...}}>
          Description
        </Label>
        <Input type=\\"textarea\\" id=\\"description\\" value=\\"\\" onChange={[Function: handleChangeText]} />
      </FormGroup>
    </ModalBody>
    <ModalFooter tag=\\"div\\">
      <Button color=\\"primary\\" type=\\"submit\\" tag=\\"button\\">
        Test button text
      </Button>
       
      <Button color=\\"secondary\\" onClick={[Function: mockConstructor]} tag=\\"button\\">
        Cancel
      </Button>
    </ModalFooter>
  </form>
</Modal>"
`;
